src/test/test-lwm2m.c

```diff
@@ -562,94 +1322,522 @@ delete_cb(void *data,
+            server_type, sol_lwm2m_client_info_get_name(cinfo));
         r = sol_lwm2m_server_del_observer(server, cinfo, "/999/0/2",
-            observe_res_cb, NULL);
+            observe_res_cb, data);
         ASSERT(r == 0);
+
+        printf("DBG: '%s' ---[Delete /999/0]---> %s\n",
+            server_type, sol_lwm2m_client_info_get_name(cinfo));
         r = sol_lwm2m_server_delete_object_instance(server, cinfo, "/999/0",
-            delete_cb, NULL);
+            delete_cb, data);
+        ASSERT(r == 0);
+    } else {
+        ASSERT(1 == 2); //TIMEOUT/UNREGISTER, this must not happen!
+    }
+}
+
+#ifdef DTLS
+static void
+write_acl_unauthorized_cb(void *data,
+    struct sol_lwm2m_server *server,
+    struct sol_lwm2m_client_info *client, const char *path,
+    enum sol_coap_response_code response_code)
+{
+    int r;
+    char *server_type = data;
+
+    ASSERT(response_code == SOL_COAP_RESPONSE_CODE_UNAUTHORIZED);
+
+    printf("DBG: '%s' ---[Delete /]---> %s\n",
+        server_type, sol_lwm2m_client_info_get_name(client));
+    r = sol_lwm2m_server_delete_object_instance(server, client, "/",
+        delete_cb, data);
+    ASSERT(r == -EINVAL);
+
+    delete_cb(data, server, client, NULL, SOL_COAP_RESPONSE_CODE_UNAUTHORIZED);
+}
+
+static void
+read_unauthorized_cb(void *data,
+    struct sol_lwm2m_server *server,
+    struct sol_lwm2m_client_info *client,
+    const char *path,
+    enum sol_coap_response_code response_code,
+    enum sol_lwm2m_content_type content_type,
+    struct sol_str_slice content)
+{
+    ASSERT(response_code == SOL_COAP_RESPONSE_CODE_UNAUTHORIZED);
+}
+
+static void
+observe_unauthorized_cb(void *data,
+    struct sol_lwm2m_server *server,
+    struct sol_lwm2m_client_info *client,
+    const char *path,
+    enum sol_coap_response_code response_code,
+    enum sol_lwm2m_content_type content_type,
+    struct sol_str_slice content)
+{
+    ASSERT(response_code == SOL_COAP_RESPONSE_CODE_UNAUTHORIZED);
+
+}
+
+static void
+sec_registration_event_cb(void *data, struct sol_lwm2m_server *server,
+    struct sol_lwm2m_client_info *cinfo,
+    enum sol_lwm2m_registration_event event)
+{
+    int r;
+    uint16_t i;
+    char *server_type = data;
+    struct sol_lwm2m_resource res[2];
+
+    if (event == SOL_LWM2M_REGISTRATION_EVENT_REGISTER) {
+        check_cinfo(cinfo, "cli1", NULL, NULL, SOL_LWM2M_BINDING_MODE_U, true);
+
+        SOL_LWM2M_RESOURCE_SINGLE_INIT(r, &res[0], DUMMY_OBJECT_BOOLEAN_FALSE_ID,
+            SOL_LWM2M_RESOURCE_DATA_TYPE_BOOL, false);
+        ASSERT(r == 0);
+        SOL_LWM2M_RESOURCE_SINGLE_INIT(r, &res[1], DUMMY_OBJECT_FLOAT_ID,
+            SOL_LWM2M_RESOURCE_DATA_TYPE_FLOAT, FLOAT_VALUE);
+        ASSERT(r == 0);
+
+        printf("DBG: '%s' ---[Write /999]---> '%s'\n",
+            server_type, sol_lwm2m_client_info_get_name(cinfo));
+        r = sol_lwm2m_server_write(server, cinfo, "/999",
+            res, sol_util_array_size(res), NULL, data);
```

Comment:
Because it must fail without a CoAP Request being even sent to the Client. Since the callback would never be called anyway (there's no Request -> there's no Response), we save code.